bokeh/util/dataclasses.py

```diff
@@ -0,0 +1,82 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2012 - 2021, Anaconda, Inc., and Bokeh Contributors.
+# All rights reserved.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+''' Provides helper function for dealing with dataclasses.
+
+'''
+
+#-----------------------------------------------------------------------------
+# Boilerplate
+#-----------------------------------------------------------------------------
+from __future__ import annotations
+
+import logging # isort:skip
+log = logging.getLogger(__name__)
+
+#-----------------------------------------------------------------------------
+# Imports
+#-----------------------------------------------------------------------------
+
+# Standard library imports
+from dataclasses import (
+    Field,
+    dataclass,
+    fields,
+    is_dataclass,
+)
+from typing import (
+    Any,
+    Iterable,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+#-----------------------------------------------------------------------------
+# Globals and constants
+#-----------------------------------------------------------------------------
+
+__all__ = (
+    "NotRequired",
+    "Unspecified",
+    "dataclass",
+    "values",
+)
+
+#-----------------------------------------------------------------------------
+# General API
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Dev API
+#-----------------------------------------------------------------------------
+
+class _UnspecifiedType:
+    def __repr__(self) -> str:
+        return "Unspecified"
+
+Unspecified = _UnspecifiedType()
+
+_T = TypeVar("_T")
+NotRequired = Union[_UnspecifiedType, _T]
+
+def values(obj: Any) -> Iterable[Tuple[Field[Any], Any]]:
```

Comment:
Note, there's `dataclasses.asdict()`, but it works recursively and more importantly it deep copies (!) leaf values.